server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -399,80 +404,104 @@ private void construct(Environment initialEnvironment, NodeServiceProvider servi
             (e, apmConfig) -> logger.error("failed to delete temporary APM config file [{}], reason: [{}]", apmConfig, e.getMessage())
         );
 
-        pluginsService = serviceProvider.newPluginService(initialEnvironment, environmentSettings);
-        final Settings settings = Node.mergePluginSettings(pluginsService.pluginMap(), environmentSettings);
+        pluginsService = serviceProvider.newPluginService(initialEnvironment, envSettings);
+        Settings settings = Node.mergePluginSettings(pluginsService.pluginMap(), envSettings);
 
         /*
          * Create the environment based on the finalized view of the settings. This is to ensure that components get the same setting
          * values, no matter they ask for them from.
          */
-        this.environment = new Environment(settings, initialEnvironment.configFile());
-        Environment.assertEquivalent(initialEnvironment, this.environment);
+        environment = new Environment(settings, initialEnvironment.configFile());
+        Environment.assertEquivalent(initialEnvironment, environment);
 
-        final List<ExecutorBuilder<?>> executorBuilders = pluginsService.flatMap(p -> p.getExecutorBuilders(settings)).toList();
+        return settings;
+    }
 
-        final ThreadPool threadPool = new ThreadPool(settings, executorBuilders.toArray(new ExecutorBuilder<?>[0]));
+    private ThreadPool createThreadPool(Settings settings) throws IOException {
+        ThreadPool threadPool = new ThreadPool(
+            settings,
+            pluginsService.flatMap(p -> p.getExecutorBuilders(settings)).toArray(ExecutorBuilder<?>[]::new)
+        );
         resourcesToClose.add(() -> ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS));
-        final ResourceWatcherService resourceWatcherService = new ResourceWatcherService(settings, threadPool);
```

Comment:
This constructor schedules some tasks - do these explicitly need to be without the HeaderContext set below?

Reply:
It's here from https://github.com/elastic/elasticsearch/pull/54993, looks like it was just moved as early as possible